File: yizhi/elytracollectutils/pathfinder/utils/Hud.java
package com.zychen027.meteorplusplus.modules.elytracollectutils.pathfinder.utils;

import net.elytraautopilot.ElytraAutoPilot;
import net.elytraautopilot.config.ModConfig;
import net.minecraft.client.MinecraftClient;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.text.Text;
import net.minecraft.util.Formatting;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;

import java.util.ArrayList;
import java.util.List;

import static net.elytraautopilot.ElytraAutoPilot.*;

public class Hud {
    private static List<Double> velocityList = new ArrayList<>();
    private static List<Double> velocityListHorizontal = new ArrayList<>();
    private static int _tick;
    private static int _index = -1;
    public static Text[] hudString;

    public static void tick() {
        _tick++;
    }

    public static void drawHud(PlayerEntity player) {
        // If GUI is disabled, clear everything
        if (!ModConfig.INSTANCE.showGui) {
            hudString = new Text[0];
            return;
        }

        double altitude = player.getPos().y;
        int gticks = Math.max(1, ModConfig.INSTANCE.groundCheckTicks);

        if (_tick >= gticks) {
            _index++;
            if (_index >= 1200 / gticks) _index = 0;
            if (velocityList.size() < 1200 / gticks) {
                velocityList.add(currentVelocity);
                velocityListHorizontal.add(currentVelocityHorizontal);
            } else {
                velocityList.set(_index, currentVelocity);
                velocityListHorizontal.set(_index, currentVelocityHorizontal);
            }
            World world = player.getWorld();
            int l = world.getBottomY();
            Vec3d clientPos = player.getPos();
            for (double i = clientPos.getY(); i > l; i--) {
                BlockPos blockPos = BlockPos.ofFloored(clientPos.getX(), i, clientPos.getZ());
                if (world.getBlockState(blockPos).isSolidBlock(world, blockPos)) {
                    groundheight = clientPos.getY() - i;
                    break;
                } else {
                    groundheight = clientPos.getY();
                }
            }
            _tick = 0;

            // Compute averages if we have enough samples
            double avgVelocity = 0, avgHorizontalVelocity = 0;
            if (velocityList.size() >= 10) {
                avgVelocity = velocityList.stream().mapToDouble(d -> d).average().orElse(0.0);
                avgHorizontalVelocity = velocityListHorizontal.stream().mapToDouble(d -> d).average().orElse(0.0);
            }

            // Build up a dynamic list of lines
            List<Text> lines = new ArrayList<>();

            if (ModConfig.INSTANCE.showEnabled) {
                lines.add(
                        Text.translatable("text.pathfinder.hud.toggleAutoFlight")
                                .append(
                                        Text.translatable(autoFlight
                                                ? "text.pathfinder.hud.true"
                                                : "text.pathfinder.hud.false"
                                        ).formatted(autoFlight ? Formatting.GREEN : Formatting.RED)
                                )
                );
            }

            if (ModConfig.INSTANCE.showAltitude) {
                lines.add(
                        Text.translatable("text.pathfinder.hud.altitude", String.format("%.2f", altitude))
                                .formatted(Formatting.AQUA)
                );
            }

            if (ModConfig.INSTANCE.showHeight) {
                String heightStr = groundheight < 0 ? "???" : String.valueOf(Math.round(groundheight));
                lines.add(
                        Text.translatable("text.pathfinder.hud.heightFromGround", heightStr)
                                .formatted(Formatting.AQUA)
                );
            }

            if (ModConfig.INSTANCE.showHeightReq) {
                boolean ready = groundheight > ModConfig.INSTANCE.minHeight;
                String req = ready
                        ? "Ready"
                        : String.valueOf(Math.round(ModConfig.INSTANCE.minHeight - groundheight));
                lines.add(
                        Text.translatable("text.pathfinder.hud.neededHeight")
                                .formatted(Formatting.AQUA)
                                .append(Text.literal(req).formatted(ready ? Formatting.GREEN : Formatting.RED))
                );
            }

            if (ModConfig.INSTANCE.showSpeed) {
                lines.add(
                        Text.translatable("text.pathfinder.hud.speed", String.format("%.2f", currentVelocity * 20))
                                .formatted(Formatting.YELLOW)
                );
            }

            if (ModConfig.INSTANCE.showAvgSpeed) {
                if (avgVelocity == 0) {
                    lines.add(
                            Text.translatable("text.pathfinder.hud.calculating")
                                    .formatted(Formatting.WHITE)
                    );
                } else {
                    lines.add(
                            Text.translatable("text.pathfinder.hud.avgSpeed", String.format("%.2f", avgVelocity * 20))
                                    .formatted(Formatting.YELLOW)
                    );
                }
            }

            if (ModConfig.INSTANCE.showHorizontalSpeed && avgVelocity != 0) {
                lines.add(
                        Text.translatable("text.pathfinder.hud.avgHSpeed", String.format("%.2f", avgHorizontalVelocity * 20))
                                .formatted(Formatting.YELLOW)
                );
            }

            // Fly-to / landing lines (you can also gate these with new config flags if you like)
            if (isflytoActive && !forceLand) {
                if (ModConfig.INSTANCE.showFlyTo) {
                    lines.add(
                            Text.translatable("text.pathfinder.flyto", argXpos, argZpos)
                                    .formatted(Formatting.LIGHT_PURPLE)
                    );
                }
                if (distance != 0 && ModConfig.INSTANCE.showEta) {
                    lines.add(
                            Text.translatable("text.pathfinder.hud.eta",
                                    String.valueOf(Math.round(distance / (avgHorizontalVelocity * 20)))
                            ).formatted(Formatting.LIGHT_PURPLE)
                    );
                }
                if (ModConfig.INSTANCE.showAutoLand) {
                    lines.add(
                            Text.translatable("text.pathfinder.hud.autoLand")
                                    .formatted(Formatting.LIGHT_PURPLE)
                                    .append(
                                            Text.translatable(ModConfig.INSTANCE.autoLanding
                                                    ? "text.pathfinder.hud.enabled"
                                                    : "text.pathfinder.hud.disabled"
                                            ).formatted(ModConfig.INSTANCE.autoLanding ? Formatting.GREEN : Formatting.RED)
                                    )
                    );
                }
                if (isLanding && ModConfig.INSTANCE.showLandingStatus) {
                    lines.add(
                            Text.translatable("text.pathfinder.hud.landing")
                                    .formatted(Formatting.LIGHT_PURPLE)
                    );
                }
            }

            if (forceLand && ModConfig.INSTANCE.showLandingStatus) {
                // Override or add a “forced landing” indicator
                lines.add(
                        Text.translatable("text.pathfinder.hud.landing")
                                .formatted(Formatting.LIGHT_PURPLE)
                );
            }

            // Finally, turn the list into your array
            hudString = lines.toArray(new Text[0]);
        }
    }


    public static void clearHud() {
        velocityList.clear();
        velocityListHorizontal.clear();
    }
}

File: src/main/kotlin/com/zychen027/MeteorPlusPlus/modules/elytraautopilot/hud/Hud.kt
package com.zychen027.meteorplusplus.modules.elytraautopilot.hud

import com.zychen027.meteorplusplus.modules.elytraautopilot.ElytraAutoPilot
import net.minecraft.client.MinecraftClient
import net.minecraft.client.font.TextRenderer
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.network.ClientPlayerEntity

/**
 * HUD 显示 - 来自 ElytraAutoPilot
 * 适配 MC 1.21.8 Matrix3x2fStack
 */
object Hud {
    private var tickCounter = 0

    /**
     * HUD tick
     */
    fun tick() {
        tickCounter++
    }

    /**
     * 绘制 HUD
     */
    fun drawHud(context: DrawContext, textRenderer: TextRenderer, player: ClientPlayerEntity) {
        val module = ElytraAutoPilot.INSTANCE ?: return
        
        if (!module.showGui.get()) return

        val lines = mutableListOf<String>()

        // 启用状态
        if (module.showEnabled.get()) {
            val status = when {
                ElytraAutoPilot.autoFlight -> "§a 自动飞行中"
                ElytraAutoPilot.onTakeoff -> "§e 起飞中"
                else -> "§c 已禁用"
            }
            lines.add("状态：$status")
        }

        // 高度
        if (module.showAltitude.get()) {
            lines.add("高度：${player.y.toInt()}")
        }

        // 高度要求
        if (module.showHeightReq.get()) {
            lines.add("最低高度：${module.minHeight.get()}")
        }

        // 速度
        if (module.showSpeed.get()) {
            val speed = String.format("%.2f", ElytraAutoPilot.currentVelocity)
            lines.add("速度：$speed")
        }

        // 平均速度
        if (module.showAvgSpeed.get()) {
            val avgSpeed = String.format("%.2f", (ElytraAutoPilot.velHigh + ElytraAutoPilot.velLow) / 2)
            lines.add("平均速度：$avgSpeed")
        }

        // 水平速度
        if (module.showHorizontalSpeed.get()) {
            val hSpeed = String.format("%.2f", ElytraAutoPilot.currentVelocityHorizontal)
            lines.add("水平速度：$hSpeed")
        }

        // 飞行目标
        if (module.showFlyTo.get() && ElytraAutoPilot.isflytoActive) {
            lines.add("目标：x=${ElytraAutoPilot.argXpos}, z=${ElytraAutoPilot.argZpos}")
        }

        // 预计到达时间
        if (module.showEta.get() && ElytraAutoPilot.isflytoActive) {
            val distance = ElytraAutoPilot.distance
            val eta = if (ElytraAutoPilot.currentVelocity > 0.01) {
                String.format("%.1f", distance / ElytraAutoPilot.currentVelocity / 20)
            } else "∞"
            lines.add("预计时间：${eta}s")
        }

        // 自动降落
        if (module.showAutoLand.get()) {
            val landStatus = when {
                ElytraAutoPilot.forceLand -> "§c 强制降落"
                ElytraAutoPilot.isLanding -> "§e 降落中"
                else -> "§a 就绪"
            }
            lines.add("降落：$landStatus")
        }

        // 绘制 - 适配 MC 1.21.8 Matrix3x2fStack
        val x = module.guiX.get()
        val y = module.guiY.get()
        val scale = module.guiScale.get() / 100.0f

        val matrices = context.matrices
        matrices.pushMatrix()
        matrices.scale(scale, scale)

        for ((i, line) in lines.withIndex()) {
            context.drawTextWithShadow(
                textRenderer,
                line,
                (x / scale).toInt(),
                (y / scale).toInt() + i * 10,
                0xFFFFFF
            )
        }

        matrices.popMatrix()
    }
}
